aiounifi/dpi.py

```diff
@@ -0,0 +1,76 @@
+"""DPI Restrictions as part of a UniFi network."""
+from typing import List
+
+from .api import APIItem, APIItems
+
+GROUP_URL = "/rest/dpigroup"  # List DPI Group configuration
+APP_URL = "/rest/dpiapp"  # List DPI App configuration
+
+
+class DPIRestrictionGroup(APIItem):
+    """Represents a DPI App configuration."""
+
+    @property
+    def id(self) -> str:
+        return self.raw["_id"]
+
+    @property
+    def attr_no_delete(self) -> bool:
+        return self.raw.get("attr_no_delete", False)
+
+    @property
+    def attr_hidden_id(self) -> str:
+        return self.raw.get("attr_hidden_id", "")
+
+    @property
+    def name(self) -> str:
+        return self.raw["name"]
+
+    @property
+    def site_id(self) -> str:
+        return self.raw["site_id"]
+
+    @property
+    def dpiapp_ids(self) -> List[str]:
+        return self.raw.get("dpiapp_ids", [])
+
+    @property
+    def enabled(self) -> bool:
+        return self.raw["enabled"]
+
+
+class DPIRestrictionGroups(APIItems):
+    """Represents DPI Group configurations."""
+
+    KEY = "name"
+
+    def __init__(self, raw: dict, request) -> None:
+        super().__init__(raw, request, GROUP_URL, DPIRestrictionGroup)
+
+    async def update(self) -> None:
+        raw_apps = await self._request("get", APP_URL)
+        raw_groups = await self._request("get", GROUP_URL)
+        for item in raw_groups:
+            item["enabled"] = all(
+                [
+                    app["enabled"]
+                    for app in raw_apps
+                    if app["_id"] in item.get("dpiapp_ids", [])
+                ]
+            )
+
+        self.process_raw(raw_groups)
+
+    async def async_enable(self, dpi: DPIRestrictionGroup) -> None:
+        """Enable DPI Restriction Group Apps."""
+        for app_id in dpi.dpiapp_ids:
+            app_url = f"{APP_URL}/{app_id}"
+            data = {"enabled": True}
+            await self._request("put", app_url, json=data)
+
+    async def async_disable(self, dpi: DPIRestrictionGroup) -> None:
+        """Disable DPI Restriction Group Apps."""
+        for app_id in dpi.dpiapp_ids:
+            app_url = f"{APP_URL}/{app_id}"
+            data = {"enabled": False}
+            await self._request("put", app_url, json=data)
```

Comment:
is it the expected way to manage all apps per group?

If yes, can we do a gather on the calls?
```suggestion
    async def async_enable(self, dpi: DPIRestrictionGroup) -> None:
        """Enable DPI Restriction Group Apps."""
        calls = []
        for app_id in dpi.dpiapp_ids:
            app_url = f"{APP_URL}/{app_id}"
            data = {"enabled": True}
            calls.append(self._request("put", app_url, json=data))
        await asyncio.gather(*calls)

    async def async_disable(self, dpi: DPIRestrictionGroup) -> None:
        """Disable DPI Restriction Group Apps."""
        calls = []
        for app_id in dpi.dpiapp_ids:
            app_url = f"{APP_URL}/{app_id}"
            data = {"enabled": False}
            calls.append(self._request("put", app_url, json=data))
        await asyncio.gather(*calls)
```

Reply:
The frontend implementation is a little strange in the Controller currently. You can't enable/disable a group, but there's no reason to have more than one app per group because the categories are an array. I guess you could enable/disable a specific set of categories independently, but there is no "name" attribute on the apps, so I went with a simplistic implementation which mandates control of all apps in a group being the same.

Reply:
Fair enough 👍🏻